scripts/generate-docs.go

```diff
@@ -0,0 +1,307 @@
+}
+
+// getBaseDir returns the terraform repo directory.
+//
+// TODO: simplify this logic with something like:
+// git rev-parse --show-toplevel
+func getBaseDir() string {
+	_, scriptPath, _, ok := runtime.Caller(0)
+	if !ok {
+		log.Fatal("Could not get current working directory")
+	}
+	for !strings.HasPrefix(filepath.Base(scriptPath), "terraform-provider-") && scriptPath != "/" {
+		scriptPath = filepath.Clean(scriptPath + "/..")
+	}
+	if scriptPath == "/" {
+		log.Fatal("Script was run outside of fastly provider directory")
+	}
+	return scriptPath
+}
+
+// getTemplate walks the templates directory filtering non-tmpl extension
+// files, and parsing all the templates found (ensuring they must parse).
+func getTemplate(tmplDir string) *template.Template {
+	var templateFiles []string
+	filepath.Walk(tmplDir, func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) == ".tmpl" {
+			templateFiles = append(templateFiles, path)
+		}
+		return nil
+	})
+	return template.Must(template.ParseFiles(templateFiles...))
+}
+
+// renderPages iterates over the given pages and renders each element.
+func renderPages(t *template.Template, pages []Page) {
+	for _, p := range pages {
+		renderPage(t, p)
+	}
+}
+
+// renderPage creates a new file based on the page information given, and
+// renders the associated template for that page.
+func renderPage(t *template.Template, p Page) {
+	basePath := filepath.Dir(p.path)
+	err := makeDirectoryIfNotExists(basePath)
+	if err != nil {
+		log.Fatal()
+	}
+
+	f, err := os.Create(p.path)
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+	err = t.ExecuteTemplate(f, p.name, p)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// makeDirectoryIfNotExists asserts whether a directory exists and makes it
+// if not. Returns nil if exists or successfully made.
+func makeDirectoryIfNotExists(path string) error {
+	fi, err := os.Stat(path)
+	switch {
+	case err == nil && fi.IsDir():
+		return nil
+	case err == nil && !fi.IsDir():
+		return fmt.Errorf("%s already exists as a regular file", path)
+	case os.IsNotExist(err):
+		return os.MkdirAll(path, 0750)
+	case err != nil:
+		return err
+	}
+
+	return nil
+}
+
+// appendSyntaxToFiles walks the temporary directory finding all the rendered
+// Markdown files we generated and proceeds to append the required template
+// syntax that the tfplugindocs tool needs.
+func appendSyntaxToFiles(tempDir string) {
+	filepath.Walk(tempDir, func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) == ".tmpl" {
+			// open file for appending and in write-only mode
+			f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0644)
+			if err != nil {
+				log.Fatal(err)
+			}
+			if _, err := f.Write([]byte("{{ .SchemaMarkdown | trimspace }}\n")); err != nil {
+				f.Close() // ignore error; Write error takes precedence
+				log.Fatal(err)
+			}
+			if err := f.Close(); err != nil {
+				log.Fatal(err)
+			}
+		}
+		return nil
+	})
+}
```

Comment:
An alternative to injecting this at the bottom of every outputed file, is to have this inline within our template templates wrapped in a `println` template function so our precomlie templating step doesn't execute the statement. I.e. add something like this to the files:

```
{{ println "{{ .SchemaMarkdown | trimspace }}" }}
```

This would give us more fine grain control over where in the file the schema appears. Those is slightly more error prone if we forget to include in a flile.

Reply:
So although I like the idea of having control over where the schema exists in the file, I think I prefer the current implementation for a couple of reasons...

1. consistency across pages (schema appears at the bottom always rather than _potentially_ changing on a per page basis)
2. it seems like for the most part the original 'reference' sections we had (which the schema section is replacing) would appear last any way.
3. less error prone as you have already stated
4. decouples the `tfplugindocs` specific syntax (e.g. if we ever needed to migrate to a different tool)